### PR TITLE
Add support for ChannelShuffle node.

### LINKED ISSUE #2722: Add ChannelShuffle node
Lowering in Lower has few advantages:
1. C2 Importer seems not to be a good place to do that.
2.  Avoid reconstructing ChannelShuffle parameters from ReshapeNode, in GraphOptimizer. In addition, GraphOptimizer just wants to swap Transpose and Channel Shuffle, so it should not have to know the Channel Shuffle operator details.
3. Optimizing Reshape->Transpose->Reshape into ChannelShuffle in the Backend is possible, but can be considered as dependent on a sequence that might change.

### PR DESCRIPTION
**Pull request by vuzelac-cadence**
*Description*:
Add ChannelShuffle node. It's lowered in Lower.
*Testing*:
Related GraphOptimizer test is updated.

Fixes #2722

### CODE REVIEW

**Comment by rdzhabarov** [comment]
Even though this adds an additional node to the node gen, this makes some of the optimizations much cleaner.

@vuzelac-cadence what was the main motivation for the change? Is that because this could be directly supported efficiently on hardware?

**Comment by vuzelac-cadence** [comment]
We have an efficient implementation of ChannelShuffle that's a bit different than Transpose. In addition, we just can't have Reshapes around as they are limiting our optimization efforts.
But, the change is done in Glow, rather than in our Backend, just because I though recognizing sequences likes Reshape->Transpose->Reshape is a bit too patchy. Not recognizing per se, but the need to reconstruct group/kernel, that we would have to do just like GraphOptimizer was doing.

**Comment by vuzelac-cadence** [comment]
continuous-integration/travis-ci/pr failure:
```
CMake Error at CMakeLists.txt:97 (message):
  LLVM minimum required version is 7.  LLVM package version found: .
```

It would be good to have ability to manually trigger CI. See #2709

**Comment by rdzhabarov** [comment]
Can you trigger CI from the circleci page? (there is rebuild option, i just triggered)

**Comment by rdzhabarov** [comment]
> rather than in our Backend, just because I though recognizing sequences likes Reshape->Transpose->Reshape is a bit too patchy

That sequence appeared on the shufflenet, iirc.

I think having separate node would make sense and let backend decide whether lower or not.
I'd keep the previous optimization in place and add the new one (sinking transpose below channelshuffle node).

**Comment by jfix71** [comment]
>I'd keep the previous optimization in place and add the new one (sinking transpose below channelshuffle node).

I think this actually would be a great fit for the `fold()` pass. We fold it back to the Node, and then can let the GraphOptimizer focus on optimizing it at a higher level, then later on we lower it again.

**Comment by vuzelac-cadence** [comment]
> Can you trigger CI from the circleci page? (there is "rebuild/rerun workflow" option, i just triggered)

That button is shaded out, saying that I don't have write permissions. Above the button is my Github Avatar and on the left side says Pytorch (should be the organization). My github is authorized with CI.

**Comment by vuzelac-cadence** [comment]
> > rather than in our Backend, just because I though recognizing sequences likes Reshape->Transpose->Reshape is a bit too patchy
> 
> That sequence appeared on the shufflenet, iirc.

Don't see in the original paper. But, since the channel shuffle is exactly that, my only problem with it was the reconstruction of the node in GraphOptimizer - just should not be there.

**Comment by vuzelac-cadence** [comment]
> > I'd keep the previous optimization in place and add the new one (sinking transpose below channelshuffle node).
> 
> I think this actually would be a great fit for the `fold()` pass. We fold it back to the Node, and then can let the GraphOptimizer focus on optimizing it at a higher level, then later on we lower it again.

Want to make sure what exactly is to be done here: 
GraphOptimizer.cpp - understood - just add the original optimization along with the new one.
GraphOptzTest - original code can't be added as it was just calling `F->createChannelShuffle()` . Seems it needs a new sequence (RN->TR->RN) that would trigger the existing optimization ?

**Comment by jfix71** [comment]
>GraphOptimizer.cpp - understood - just add the original optimization along with the new one.

Actually I think we just should keep the new optimization and drop the original. The old logic was basically to get around the issue of not having the unlowered version, and can be moved/refactored into `fold()`. Because we will fold it, I don't think we need to have both paths in the GraphOptimizer.

>GraphOptzTest - original code can't be added as it was just calling F->createChannelShuffle() . Seems it needs a new sequence (RN->TR->RN) that would trigger the existing optimization ?

I think the change you made that verifies that the new version of the optimization works is good. I think you should also add a `GraphFold` test (see the couple that already exist in `GraphOptzTest.cpp`) that checks that the sequence we are going to fold into the `ChannelShuffleNode` is in fact correctly folded.

**Comment by vuzelac-cadence** [comment]
> > GraphOptimizer.cpp - understood - just add the original optimization along with the new one.
> 
> Actually I think we just should keep the new optimization and drop the original. The old logic was basically to get around the issue of not having the unlowered version, and can be moved/refactored into `fold()`. Because we will fold it, I don't think we need to have both paths in the GraphOptimizer.

Something is wrong here. I am not looking for more work but still want to avoid breaking anything - comment in GraphTest says you encountered Reshape->Transpose->Reshape in ShuffleNet. If so, removing this optimization would break your ShuffleNet, no ? (hopefully not :) ).  Curious where that ShuffleNet comes from ? TensorFlow one for sure will include RN->TR->RN as both TF and ONNX don't support ChannelShuffle. But that model should fail to compile in Glow due to other problems before it could even fail here.

**Comment by vuzelac-cadence** [comment]
> I think the change you made that verifies that the new version of the optimization works is good. I think you should also add a `GraphFold` test (see the couple that already exist in `GraphOptzTest.cpp`) that checks that the sequence we are going to fold into the `ChannelShuffleNode` is in fact correctly folded.

I didn't get this one - that test calls `glow::fold() `which calls optimizeReshape() and foldLeakyRelu() only - it'll not fold `ChannelShuffle`. But also, I am not sure how the existing test is not sufficient: sinkTransposeBelowChannelShuffleNodesAndEliminate does seem to trigger the folding and checks if it's done correctly (?).

**Comment by vuzelac-cadence** [comment]
Failing CI test (fails in OperatorTest) seems not to be related to my change:

`*** Error in `/var/lib/jenkins/workspace/build/tests/OperatorTest': double free or corruption (!prev): 0x0000000003f45ef0 ***`


`

**Comment by jfix71** [comment]
@vuzelac-cadence Right now we have the following situation. Our Caffe2 Shufflenet sees it is importing a `ChannelShuffle` and calls `createChannelShuffle()` which creates Reshape-Transpose-Reshape (i.e. it's lowered directly at import time). This is why we later have the graph optimization pass `sinkTranposeBelowChannelShuffle()` which looks for this pattern along with a transpose above it and sinks the transpose below it. Changing `createChannelShuffle()` to create a `ChannelShuffleNode` means this optimization pass is no longer necessary, given the new version in this PR.

What I am suggesting for the new version with the newly added `ChannelShuffleNode` is to add a new pass to `fold()`, which is similar to the logic in `sinkTranposeBelowChannelShuffle()`, except it does no transpose sinking. It still looks for Reshape-Transpose-Reshape, but then simply and replaces it by the `ChannelShuffleNode` you are introducing. This fold pass was recently introduced by @tlepley-cadence. It occurs before any optimizations or transformations to the graph. Its purpose is to make analysis and optimizations of graphs much easier, for purposes exactly like this, where we might load graphs that have pre-lowered sequences and we'd prefer to see them unlowered.

>TensorFlow one for sure will include RN->TR->RN as both TF and ONNX don't support ChannelShuffle.

This should solve the issue for TF/ONNX, as the RN->TR->RN sequence will be folded into a `ChannelShuffleNode` before any optimizations/transformations. Then the `sinkTranposeBelowChannelShuffle()` pass will work correctly as implemented already in this PR. Note that we don't need this folding logic for our Caffe2 Shufflenet after this PR, as we will directly create the `ChannelShuffleNode` when importing -- it's only beneficial for the ONNX/TF version.

>I didn't get this one - that test calls `glow::fold() `which calls `optimizeReshape()` and `foldLeakyRelu() `only - it'll not fold `ChannelShuffle`. But also, I am not sure how the existing test is not sufficient: `sinkTransposeBelowChannelShuffleNodesAndEliminate `does seem to trigger the folding and checks if it's done correctly (?).

I'm suggesting you add a new `foldChannelShuffle()` pass to `fold()`, as I described above, looking for RN->TR->RN and replacing it with a `ChannelShuffleNode`. The `sinkTransposeBelowChannelShuffleNodesAndEliminate` test is totally find to keep as well. I'm saying when you add the new `fold()` logic you should add a test to verify it correctly folds a RN->TR->RN sequence into a `ChannelShuffleNode`. It should be pretty simple to create.

**Comment by jackm321** [comment]
As the author of the original `sinkTranposeBelowChannelShuffle()` I 100% agree with @jfix71 that we should kill it and replace it by folding Reshape-Transpose-Reshape into a `ChannelShuffleNode` and using the logic from this PR instead.

File: tests/unittests/GraphOptzTest.cpp [hunk 1]
```diff
@@ -2089,6 +2089,32 @@ TEST_F(GraphFold, foldLeakyReluFromConst) {
...
+  ::glow::fold(F_, CompilationMode::Infer);
+
+  ASSERT_EQ(F_->getNodes().size(), 2);
+
+  // Check for ChannelShuffle node.
+  auto *CS = llvm::dyn_cast<ChannelShuffleNode>(save->getInput().getNode());
+  ASSERT_NE(nullptr, CS);
+
+  // Ensure ChannelShuffle node has the same dimensions as the input.
+  EXPECT_EQ(CS->getType(0)->dims(), llvm::makeArrayRef(inputDims));
```
Comment (jfix71) [inline tests/unittests/GraphOptzTest.cpp]: nit: `CS->getResult().dims()` here

Comment (jfix71) [reply tests/unittests/GraphOptzTest.cpp]: Can you also check that group and kernel are set as expected?

File: tests/unittests/GraphOptzTest.cpp [hunk 2]
```diff
@@ -2707,37 +2733,20 @@ TEST_F(GraphOptz, sinkTransposeBelowChannelShuffleNodesAndEliminate) {
...
-  EXPECT_EQ(preShuffleRN->getDims(),
-            llvm::makeArrayRef(expectedPreShuffleRNDims));
+  ASSERT_EQ(F_->getNodes().size(), 2);
+
+  // Check that the channel shuffle node is still there.
+  auto *CSN = llvm::dyn_cast<ChannelShuffleNode>(save->getInput().getNode());
+  ASSERT_NE(nullptr, CSN);
+
+  // Ensure ChannelShuffle node has the same dimensions as the input.
+  EXPECT_EQ(CSN->getType(0)->dims(), llvm::makeArrayRef(inputDims));
```
Comment (jfix71) [inline tests/unittests/GraphOptzTest.cpp]: Same as in other test: `CS->getResult().dims()` here, and check that group and kernel are set as expected?

File: lib/Optimizer/Lower.cpp [hunk 3]
```diff
@@ -641,6 +641,36 @@ static void lowerTileNode(Function *F, LoweredInfoMap *loweredMap,
...
+  assert(D % group == 0);
+
+  dims.erase(dims.begin() + kernel);
+  // Reshape {D1, ... D_k, ... D_n} -> {D1, ... group, D_k / group, ... D_n}
+  dims.insert(dims.begin() + kernel, D / group);
+  dims.insert(dims.begin() + kernel, group);
+  auto *R1 = F->createReshape(CSN.getName().str() + ".reshape1", input, dims);
+
+  std::vector<unsigned_t> transpose(dims.size());
+  for (size_t i = 0; i < transpose.size(); i++)
```
Comment (jfix71) [inline lib/Optimizer/Lower.cpp]: nit: I see this was copied from before, but we prefer curly brackets even on single line nests.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 4]
```diff
@@ -2512,6 +2445,87 @@ static void foldLeakyRelu(Function *F) {
...
+  return resM;
+}
+
+// Fold Reshape->Transpose->Reshape into ChannelShuffle when applicable.
+static void foldChannelShuffle(Function *F) {
+
+  auto &nodes = F->getNodes();
+  for (auto &node : nodes) {
+    auto *RN2 = dyn_cast<ReshapeNode>(&node);
+    if (!RN2 || RN2->getDims().size() != 4) {
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer.cpp]: I don't think it's a requirement that the ChannelShuffle input has 4 dimensions, is it? I think we can drop the checks for the dims size and just let `getChannelShuffleParams()` handle the logic for checking if it's a valid ChannelShuffle.

File: lib/Optimizer/Lower.cpp (hunk 3 above)
Comment (vuzelac-cadence) [reply lib/Optimizer/Lower.cpp]: @jfix71, note that checking the kernel picked up a bug in setting kernel field in `createChannelShuffle` in  `foldChannelShuffle()`.
